### ISSUE TITLE
[ROCm] IndexerKPool calls aiter fp8_mqa_logits unbounded; >2 GiB logits abort every TP rank (#36960 does not cover this class)

### ISSUE THREAD

**Issue by Arist12** [sgl-project/sglang#37478]
## Summary

On ROCm, `IndexerKPool` (`dsa_indexer_kpool.py`) calls aiter's `fp8_mqa_logits` with **no bound at all** on the fp32 `[num_q x num_k]` logits tensor. Above 2 GiB that kernel takes a `gl.store` fallback that fails to *compile*, and the resulting LLVM assertion `abort()`s the process rather than raising — so every TP rank dies and the server is gone.

This is the same underlying kernel limit as #36960, but **#36960 does not fix it here**: that PR patches `dsa_indexer.py::Indexer`, while `IndexerKPool` is a separate class (`MultiPlatformOp` vs `BaseFusedOp`, no inheritance) with its own code path. GLM-5.3-Flash cannot avoid `IndexerKPool` — its `config.json` declares `index_kpool = 4`, so the KPool indexer is architectural rather than a flag.

`dsa_indexer_kpool.py` does define `_should_chunk_mqa_logits`, but **it has no callers** (`grep -c 'self\._should_chunk_mqa_logits('` → `0`, both at `aa8c950a3d` and at branch head `f6d4309775`), and there is no chunked implementation behind it. Even if it were wired up, it bounds only by free/total memory, which never binds first: on MI355X measured `total_mem` 309.2 GB and `free_mem` 308.6 GB, so its two rules trigger at ~92.8 GB and ~154 GB, both far above the 2.147 GB the kernel can actually address.

## Reachability

This is not a synthetic shape. With `--chunked-prefill-size 8192` the limit is `num_k = 65535`. A long agentic context passes that: a completed DeepSWE v1.1 trial on our fleet measured `peak_context_tokens = 119,958`, i.e. `8192 x 120000 x 4 = 3.93 GB` of logits.

## Reproduction

MI355X (gfx950), ROCm 7.2.4, image `rocm/sgl-dev:v0.5.18-rocm724-mi35x-20260901`, `NUM_HEADS = 32` and `head_dim = 128` (GLM-5.3-Flash's `index_n_heads` / `index_head_dim`). Each case in its own subprocess, because the failure is `abort()` and an in-process `try/except` would not see it.

Calling `aiter.ops.triton.fp8_mqa_logits.fp8_mqa_logits` directly, boundary at `sqrt(2**31 / 4) = 23170.475`:

| `num_q = num_k` | logits bytes | >= 2 GiB | result |
|---|---|---|---|
| 23168 | 2,147,024,896 | no | **returns**, shape `(23168, 23168)` |
| 23180 | 2,149,249,600 | yes | **`SIGABRT` (rc=-6)** |

The abort:

```
python3: /__w/triton/triton/llvm-project/llvm/include/llvm/ADT/Sequence.h:275:
llvm::iota_range<unsigned int>::iota_range(T, T, bool) [T = unsigned int]:
Assertion `Begin <= End && "Begin must be less or equal to End."' failed.
```

And the production-shaped case, through `IndexerKPool._fp8_mqa_logits` as written at `aa8c950a3d`:

| shape | logits | result |
|---|---|---|
| `num_q=8192, num_k=120000` | 3.93 GB | **`SIGABRT` (rc=-6)** |

## Proposed fix

Split the **query** dimension inside the HIP branch of `IndexerKPool._fp8_mqa_logits`. This is exact rather than an approximation: output row `i` depends only on `q_fp8[i]`, `weights[i]`, `starts[i]`, `ends[i]` and the whole KV, so blocking over rows and concatenating reproduces the unsplit result exactly.

```python
num_q = q_fp8.shape[0]
num_k = k_fp8.shape[0]
max_rows = (
    _MQA_LOGITS_MAX_BYTES_ROCM // (num_k * _MQA_LOGITS_BYTES_PER_ELEM)
    if num_k > 0
    else num_q
)
if 0 < max_rows < num_q:
    return torch.cat(
        [
            fp8_mqa_logits(
                q_fp8[i : i + max_rows], k_fp8, k_scale,
                weights[i : i + max_rows],
                starts[i : i + max_rows],
                ends[i : i + max_rows],
                clean_logits=clean_logits,
            )
            for i in range(0, num_q, max_rows)
        ],
        dim=0,
    )
```

Validated, same hardware:

| case | result |
|---|---|
| exactness: blocked (forced 1000-row blocks) vs unsplit, `8192 x 20000` | **bit-identical**, `max_abs_diff = 0.0` |
| `23168 x 23168` through the patched wrapper | returns |
| `23180 x 23180` through the patched wrapper | returns (was `SIGABRT`) |
| `8192 x 120000` unpatched | `SIGABRT` |
| `8192 x 120000` patched | returns, shape `(8192, 120000)` |

## Note on where this can land

I have not opened a PR because there is currently nothing on the branch for it to patch: the HIP branch of `_fp8_mqa_logits` was introduced by #36607 and then reverted in `c767511ea8` ("Revert \"[AMD] Enable GLM-5.3-Flash on gfx942 and gfx950 (#36607)\"", 2026-09-01), with no stated reason and no re-land since. At branch head `f6d4309775` the three call sites go straight to `deep_gemm.fp8_mqa_logits`, so there is no ROCm path at all.

I did not want to re-land someone else's reverted change unilaterally. **If #36607 is going to be re-landed, this bound should go in with it** — happy to open the PR against whatever branch you prefer, or to fold the patch into the re-land. The diff is 48 lines and is attached above.

Related: #36960 (same kernel limit, `Indexer` class, targets `main`), ROCm/aiter#5114 (the underlying `gl.store` compile failure).

**Comment by Arist12** [comment]
Branch with the patch applied on top of `aa8c950a3d` (#36607's merge commit), ready to open as a PR against whichever branch you prefer once #36607 re-lands:

- https://github.com/Arist12/sglang/tree/rocm-kpool-mqa-logits-2gib-bound
- diff: https://github.com/sgl-project/sglang/compare/xinyuan/glm-5.3-flash-support...Arist12:sglang:rocm-kpool-mqa-logits-2gib-bound

48 lines, single file. It applies cleanly to `aa8c950a3d`; it does not apply to current branch head only because the HIP branch it patches was removed by the revert in `c767511ea8`.

Also happy to just hand the hunk over for inclusion in the re-land if that is simpler than a separate PR.

**Comment by Arist12** [comment]
Update, now that #36960 has merged (2026-09-01T22:40:13Z): I re-checked `main` and it is clean. `dsa_indexer.py` carries `_MQA_LOGITS_MAX_BYTES_ROCM` at line 207, applies it at line 1020, and `_should_chunk_mqa_logits` has a live call site at line 1109 — so the ROCm aiter path on `main` is bounded. The two remaining `deep_gemm.fp8_mqa_logits` calls on `main` are both inside `_get_topk_ragged_with_cp` behind `_pad_heads_for_deep_gemm` (CUDA/DeepGEMM-specific), and `dsv4/indexer.py`'s call is gated by `if not is_cuda() or is_hip():`. A repo-wide code search for `kpool` on `main` returns 0 hits.

So this report is **not** a gap in #36960 — #36960 is complete for `main`. It is specific to `IndexerKPool`, which exists only on `xinyuan/glm-5.3-flash-support`, and it is latent there right now because the ROCm path was reverted in `c767511ea8`.

Worth flagging because it is easy to conflate: there are two separate reverts here.
- `c767511ea8` reverts #36607 (the **code**) — 22 files, +111/-1722, including `dsa_indexer_kpool.py` (+14/-73), `quark/quark.py` (+10/-79), `glm5_next.py` (+3/-79), and it deletes two test files. On the feature branch.
- #37380 reverts #36608 (the **recipes**) — 3 files, all under `docs/`. On `main`.

The exposure is real today for anyone following OneNexus/GLM-5.3-Flash-MXFP4's published recipe, which mounts a #36607 checkout as a source overlay — that is how it was hit here.
